### DISCUSSION TITLE
Online video game back-end using AWS Gamelift with rusoto_gamelift crate

### DISCUSSION THREAD

**Discussion by Tipnos** [actix/actix-web#1922] [Q&A] [answered]
Hello everyone,

I'm currently working on a back-end service of an online multiplayer game which use Amazon Web Service Gamelift for matchmaking and server fleet scaling. 

I'm facing some difficulty with integrating [rusoto_gamelift crate](https://crates.io/crates/rusoto_gamelift)  which provide some tools to format and send request to AWS, the problem comes from the HttpClient.

Firstly I tried to use their [default http client](https://rusoto.github.io/rusoto/rusoto_core/request/struct.HttpClient.html) : 

```rust
use rusoto_gamelift::GameLiftClient;
use rusoto_core::credential::{EnvironmentProvider};
use rusoto_core::request::HttpClient;
use rusoto_core::region::Region;

pub async fn get_gamelift_client() -> GameLiftClient {
    let cred = EnvironmentProvider::default();
    let client = HttpClient::new().unwrap();
    GameLiftClient::new_with(client, cred, Region::EuWest1)
}

-----------------------------------------------------
// main.rs
// --snip-- 
let gamelift = get_gamelift_client().await;
App::new()
            .data(gamelift.to_owned())
// --snip--
```

But I faced this error at runtime: _thread 'actix-web' panicked at 'not currently running on the Tokio runtime.', /home/tristan/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.0.2/src/runtime/blocking/pool.rs:84:33_

I implemented the solution mentioned [here](https://github.com/actix/actix-web/issues/1283) to start actix web in a tokio runtime. But sadly, I faced exactly the same error at runtime.

Then I tried to implement their [DispatchSignedRequest](https://rusoto.github.io/rusoto/rusoto_core/request/trait.DispatchSignedRequest.html) trait with actix web client. I had then the following compiler error: _`Rc<awc::ClientConfig>` cannot be sent between threads safely
within `GameliftClient`, the trait `std::marker::Send` is not implemented for `Rc<awc::ClientConfig>`_

I found already an issue [here](https://github.com/actix/actix-web/issues/1351) on this but at the moment ClientConfig is still Rc and not Arc.

I guess people have already implemented AWS services with this crate so I may miss something big here...

Any help is really appreciated! Thanks.

**Comment by fakeshadow** [comment]
Use https://crates.io/crates/rusoto_gamelift/0.45.0 for actix-web v3. (or use actix-web v4 beta with 0.46)
There is a dependency session on the right of crates.io and you want to make sure different crates have compat version of tokio. (tokio does not compat between differen major releases.)

For client like the compiler tells. actix-web client is not thread safe. It's meant to be used on single thread. So it's not possible to send it between threads. It's not an issue or bug as it's designed to do exactly single thread work.

**Reply by Tipnos** [reply]
Hi fakeshadow,

thanks for your quick and clear response. I downgrade rusoto_gamelift to match tokio compat version but I still get the same error at runtime. I checked all my others crate to see if there was an other compat problem elsewhere it seems not. Here is my Cargo.toml just in case: 

```
[package]
name = "rigidity-application"
version = "0.1.0"
authors = ["He-aD <ge.tristan@gmail.com>"]
edition = "2018"

[dependencies]
actix-identity = "0.3"
actix-rt = "1.0.0"
actix-web = { version = "3", features = ["openssl"]}
actix-files = "0.3"
actix-service="1.0.6"
actix="0.10.0"
actix-web-actors = "3.0.0"
chrono = { version = "0.4", features = ["serde"] }
derive_more = "0.99"
diesel = { version = "1.4.5", features = ["postgres","uuidv07", "r2d2", "chrono"] }
diesel-derive-enum = { version = "1.1.0", features = ["postgres"] }
dotenv = "0.15.0"
env_logger="0.8.2"
futures = "0.3"
lazy_static = "1.4"
r2d2 = "0.8.9"
rust-argon2="0.8"
serde = "1.0"
serde_derive = "1.0"
serde_json = "1.0"
sparkpost="0.5"
uuid = { version = "0.8", features = ["serde", "v4"] }
time = "0.2"
mailgun-rs="0.1.3"
rand="0.8.1"
rusoto_core = "0.43.0"
rusoto_gamelift="0.43.0"
```

I tried to upgrade to actix-web v4 beta, as you mentionned, but it looks like I have to do some refactoring to run it, which I'd like to avoid at that time. 

I also tried to wrap actix-web in a tokio runtime as mentionned in my first post with the rusoto_gamelift downgrade version, but it didn't do the trick. 

Other ideas ?

Thanks for your time.

**Comment by fakeshadow** [comment, marked as answer]
```rust
use actix_web::{get, App, HttpServer, web::Data};
use rusoto_gamelift::GameLiftClient;
use rusoto_core::credential::{EnvironmentProvider};
use rusoto_core::request::HttpClient;
use rusoto_core::region::Region;

#[actix_web::main]
async fn main() -> std::io::Result<()> {
    let cred = EnvironmentProvider::default();
    let client = HttpClient::new().unwrap();
    let client = GameLiftClient::new_with(client, cred, Region::EuWest1);

    HttpServer::new(move || App::new().data(client.clone()).service(index))
        .bind("127.0.0.1:8000")?
        .run()
        .await
}

#[get("/")]
async fn index(data: Data<GameLiftClient>) -> &'static str {
    "working!"
}

[dependencies]
actix-web = "3"
rusoto_core = "0.43.0"
rusoto_gamelift = "0.43.0"
```

**Reply by Tipnos** [reply]
Oh sorry, I see. My post wasn't enough precise, I have the runtime error when using the GameLiftClient to send a request, example: 

``` rust
// handlers.rs
pub async fn start_matchmaking(
    Path(custom_room_id): Path<i32>,
    id: Identity,
    ws: web::Data<Addr<WebsocketLobby>>,
    gamelift: web::Data<GameLiftClient>,
    pool: web::Data<Pool>
) -> AppResult<HttpResponse> {
    let user_id = id.identity().unwrap();
    match web::block(move || 
        block_on(service::start_matchmaking(
            custom_room_id,
            user_id.parse::<i32>().unwrap(),
            ws.get_ref().to_owned(),
            gamelift.get_ref(),
            &pool.get().unwrap()))).await {
        Ok(_custom_room) => {
            Ok(HttpResponse::Ok().finish())
        }
        Err(err) => match err {
            BlockingError::Error(service_error) => Err(service_error),
            BlockingError::Canceled => Err(AppError::InternalServerError(err.to_string())),
        }
    }
}

-------------------------------------------
// service.rs
pub async fn start_matchmaking(
    custom_room_id: i32,
    user_id: i32, 
    ws: Addr<WebsocketLobby>,
    gamelift: &GameLiftClient,
    conn: &PgConnection
) -> AppResult<()> {
    match custom_room::get(&custom_room_id, conn) {
        Ok(tuple) => {
            match gamelift.start_matchmaking(
                tuple.0.get_start_matchmaking_input(&tuple.1)
            ).await { // tokio panic here
// --snip-- 
```

**Reply by fakeshadow** [reply]
What is `block_on`? futures::block_on use it's own runtime which has nothing to do with either tokio or actix-web.
Your code has fundamental problem with nested web::block have async code inside.
web::block is meant for running long time sync code. There should not be any async stuff inside it.

**Reply by Tipnos** [reply]
Ok, that was it !

I added the futures::executor::block_on because of compiler mentioned it and fix an error. With your explanation I understand now why, rookie mistake :).
Thanks a lot for your time and clear answers.

Have a great week-end !
